<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>DCATools.jl Documentation · Package name</title><link href="https://cdnjs.cloudflare.com/ajax/libs/normalize/4.2.0/normalize.min.css" rel="stylesheet" type="text/css"/><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/default.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.2.0/require.min.js" data-main="assets/documenter.js"></script><script src="siteinfo.js"></script><script src="../versions.js"></script><link href="assets/documenter.css" rel="stylesheet" type="text/css"/></head><body><nav class="toc"><h1>Package name</h1><select id="version-selector" onChange="window.location.href=this.value" style="visibility: hidden"></select><form class="search" id="search-form" action="search.html"><input id="search-query" name="q" type="text" placeholder="Search docs"/></form><ul><li class="current"><a class="toctext" href="index.html">DCATools.jl Documentation</a><ul class="internal"><li><a class="toctext" href="#Data-structure-1">Data structure</a></li><li><a class="toctext" href="#Alignment-tools-–-alignmenttools.jl-1">Alignment tools – alignmenttools.jl</a></li><li><a class="toctext" href="#Input-/-output-inputoutput.jl-1">Input / output - inputoutput.jl</a></li><li><a class="toctext" href="#Model-tools-–-modeltools.jl-1">Model tools – modeltools.jl</a></li><li><a class="toctext" href="#Contact-prediction-–-contactprediction.jl-1">Contact prediction – contactprediction.jl</a></li></ul></li></ul></nav><article id="docs"><header><nav><ul><li><a href="index.html">DCATools.jl Documentation</a></li></ul><a class="edit-page" href="https://github.com/PierreBarrat/DCATools/blob/master/docs/src/index.md"><span class="fa"></span> Edit on GitHub</a></nav><hr/><div id="topbar"><span>DCATools.jl Documentation</span><a class="fa fa-bars" href="#"></a></div></header><h1><a class="nav-anchor" id="DCATools.jl-Documentation-1" href="#DCATools.jl-Documentation-1">DCATools.jl Documentation</a></h1><ul><li><a href="index.html#DCATools.jl-Documentation-1">DCATools.jl Documentation</a></li><ul><li><a href="index.html#Data-structure-1">Data structure</a></li><li><a href="index.html#Alignment-tools-–-alignmenttools.jl-1">Alignment tools – alignmenttools.jl</a></li><ul><li><a href="index.html#Compute-pairwise-frequencies.-1">Compute pairwise frequencies.</a></li><li><a href="index.html#Compute-weights-of-sequences.-1">Compute weights of sequences.</a></li></ul><li><a href="index.html#Input-/-output-inputoutput.jl-1">Input / output - inputoutput.jl</a></li><ul><li><a href="index.html#Read-DCA-parameters-from-file.-1">Read DCA parameters from file.</a></li><li><a href="index.html#Write-DCA-parameters-to-file.-1">Write DCA parameters to file.</a></li><li><a href="index.html#Read-a-numerical-alignment-from-a-file.-1">Read a numerical alignment from a file.</a></li></ul><li><a href="index.html#Model-tools-–-modeltools.jl-1">Model tools – modeltools.jl</a></li><ul><li><a href="index.html#Switch-gauge-of-DCA-parameters.-1">Switch gauge of DCA parameters.</a></li><li><a href="index.html#Compute-energies-of-sequences-for-given-DCA-parameters.-1">Compute energies of sequences for given DCA parameters.</a></li><li><a href="index.html#Infer-a-profile-model-from-frequencies-or-an-alignment.-1">Infer a profile model from frequencies or an alignment.</a></li></ul><li><a href="index.html#Contact-prediction-–-contactprediction.jl-1">Contact prediction – contactprediction.jl</a></li><ul><li><a href="index.html#Compute-Fapc-score-for-given-coupling-matrix.-1">Compute Fapc score for given coupling matrix.</a></li><li><a href="index.html#Compute-PPV-for-given-scores-and-distances.-1">Compute PPV for given scores and distances.</a></li></ul></ul></ul><h2><a class="nav-anchor" id="Data-structure-1" href="#Data-structure-1">Data structure</a></h2><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="DCATools.DCAgraph" href="#DCATools.DCAgraph"><code>DCATools.DCAgraph</code></a> — <span class="docstring-category">Type</span>.</div><div><div><pre><code class="language-none">struct DCAgraph</code></pre></div></div><a class="source-link" target="_blank" href="https://github.com/PierreBarrat/DCATools/blob/2c962efbb29e5d8f1db5365d439bae02d3b36753/src/DCATools.jl#L6-L8">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="Base.:*" href="#Base.:*"><code>Base.:*</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">function *(B, g::DCAgraph)</code></pre><p>Multiply fields and couplings in <code>g</code> by scalar <code>B</code>. Useful to change temperature. </p></div></div><a class="source-link" target="_blank" href="https://github.com/PierreBarrat/DCATools/blob/2c962efbb29e5d8f1db5365d439bae02d3b36753/src/DCATools.jl#L17-L21">source</a><div><div><pre><code class="language-none">function *(B, g::DCAgraph)</code></pre><p>Multiply fields and couplings in <code>g</code> by scalar <code>B</code>. Useful to change temperature. </p></div></div><a class="source-link" target="_blank" href="https://github.com/PierreBarrat/DCATools/blob/2c962efbb29e5d8f1db5365d439bae02d3b36753/src/DCATools.jl#L26-L30">source</a></section><h2><a class="nav-anchor" id="Alignment-tools-–-alignmenttools.jl-1" href="#Alignment-tools-–-alignmenttools.jl-1">Alignment tools – alignmenttools.jl</a></h2><h3><a class="nav-anchor" id="Compute-pairwise-frequencies.-1" href="#Compute-pairwise-frequencies.-1">Compute pairwise frequencies.</a></h3><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="DCATools.computefreqs" href="#DCATools.computefreqs"><code>DCATools.computefreqs</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">computefreqs(Y::Array{Int64,2}, w::Array{Float64,1}, q::Int64)</code></pre><p>Base routine for computing frequencies in sample <code>Y</code>. <code>w</code> is an array containing the weights. </p></div></div><a class="source-link" target="_blank" href="https://github.com/PierreBarrat/DCATools/blob/2c962efbb29e5d8f1db5365d439bae02d3b36753/src/alignmenttools.jl#L3-L7">source</a><div><div><pre><code class="language-none">computefreqs(Y::Array{Int64,2}; q = findmax(Y)[1], computew=false, weights=[], theta=0.2, saveweights=&quot;&quot;)</code></pre><p>Compute pairwise frequencies for an array input. <code>Y</code> is an array of <code>Int64</code>. Return frequencies <code>f1</code> and <code>f2</code> and weights.</p><p>Keywords: </p><ul><li><code>q</code>: default <code>findmax(Y)[1]</code></li><li><code>weights</code>: default <code>[]</code>. If it is a <code>String</code>, phylogenetic weights are read from the corresponding file. If it is an <code>Array{Float64,1}</code>, they are used directly.</li><li><code>computew</code>: default <code>false</code>. If true, phylogenetic weights are computed, calling the appropriate <code>computeweights</code>. <code>weights</code> is then ignored. </li><li><code>saveweights</code> and <code>theta</code>: see <code>computeweights</code>. </li></ul></div></div><a class="source-link" target="_blank" href="https://github.com/PierreBarrat/DCATools/blob/2c962efbb29e5d8f1db5365d439bae02d3b36753/src/alignmenttools.jl#L36-L46">source</a><div><div><pre><code class="language-none">computefreqs(msa::String; q = 0, computew=false, weights=[], theta=0.2, saveweights=&quot;&quot;, header=false, format=1)</code></pre><p>Compute pairwise frequencies for a file input. <code>msa</code> is a file containing the alignment in numerical format. Return frequencies <code>f1</code> and <code>f2</code> and weights.</p><p>Keywords: </p><ul><li><code>q</code>: default <code>findmax(Y)[1]</code></li><li><code>weights</code>: default <code>[]</code>. If it is a <code>String</code>, phylogenetic weights are read from the corresponding file. If it is an <code>Array{Float64,1}</code>, they are used directly.</li><li><code>computew</code>: default <code>false</code>. If true, phylogenetic weights are computed, calling the appropriate <code>computeweights</code>. <code>weights</code> is then ignored. </li><li><code>saveweights</code> and <code>theta</code>: see <code>computeweights</code>. </li><li><code>header</code> and <code>format</code>: see <code>readmsanum</code>.</li></ul></div></div><a class="source-link" target="_blank" href="https://github.com/PierreBarrat/DCATools/blob/2c962efbb29e5d8f1db5365d439bae02d3b36753/src/alignmenttools.jl#L79-L90">source</a></section><h3><a class="nav-anchor" id="Compute-weights-of-sequences.-1" href="#Compute-weights-of-sequences.-1">Compute weights of sequences.</a></h3><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="DCATools.computeweights" href="#DCATools.computeweights"><code>DCATools.computeweights</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">computeweights(msa::String; theta::Float64 = 0.2, saveweights::String = &quot;&quot;, format=1, header=false)</code></pre><p>Compute weights for file input. Compute weights of each sequence in file <code>msa</code> using reweighting threshold <code>theta</code> (default 0.2). </p><p>Keywords: </p><ul><li><code>theta</code>: threshold of similarity under which sequences are weighted down. Default <code>0.2</code>. </li><li><code>saveweights</code>: weights are saved there if non empty. Default <code>&quot;&quot;</code></li><li><code>format</code> and <code>header</code>: used to read <code>msa</code>. See <code>readmsanum</code>. </li></ul></div></div><a class="source-link" target="_blank" href="https://github.com/PierreBarrat/DCATools/blob/2c962efbb29e5d8f1db5365d439bae02d3b36753/src/alignmenttools.jl#L101-L110">source</a><div><div><pre><code class="language-none">computeweights(Y::Array{Int64,2} ; theta = 0.2, saveweights=&quot;&quot;)</code></pre><p>Compute weights for array input. Compute weights of each sequence in file <code>msa</code> using reweighting threshold <code>theta</code> (default 0.2). </p><p>Keywords: </p><ul><li><code>theta</code>: threshold of similarity under which sequences are weighted down. Default <code>0.2</code>. </li><li><code>saveweights</code>: weights are saved there if non empty. Default <code>&quot;&quot;</code></li></ul></div></div><a class="source-link" target="_blank" href="https://github.com/PierreBarrat/DCATools/blob/2c962efbb29e5d8f1db5365d439bae02d3b36753/src/alignmenttools.jl#L117-L125">source</a><div><div><pre><code class="language-none">computeweights(Y::Array{Int64,2},theta::Float64)</code></pre><p>Basic routine. Compute weights of sequences in alignment <code>Y</code>, using threshold <code>theta</code>. </p></div></div><a class="source-link" target="_blank" href="https://github.com/PierreBarrat/DCATools/blob/2c962efbb29e5d8f1db5365d439bae02d3b36753/src/alignmenttools.jl#L134-L138">source</a></section><h2><a class="nav-anchor" id="Input-/-output-inputoutput.jl-1" href="#Input-/-output-inputoutput.jl-1">Input / output - inputoutput.jl</a></h2><h3><a class="nav-anchor" id="Read-DCA-parameters-from-file.-1" href="#Read-DCA-parameters-from-file.-1">Read DCA parameters from file.</a></h3><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="DCATools.readparam" href="#DCATools.readparam"><code>DCATools.readparam</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">function readparam(infile::String ; format=&quot;mcmc&quot;, q=0)</code></pre><p>Read dca parameters from <code>infile</code>. Format option can be either </p><ul><li><code>&quot;mcmc&quot;</code>: <code>J i j a b val</code></li><li><code>&quot;mat&quot;</code>: One line of <code>infile</code> represents the vector <code>J[i,a][:]</code>. This is useful for parameters stored in dlm format. Optional argument <code>q</code> is needed in this case. </li></ul><p>Output is of type <code>DCAgraph</code>.</p></div></div><a class="source-link" target="_blank" href="https://github.com/PierreBarrat/DCATools/blob/2c962efbb29e5d8f1db5365d439bae02d3b36753/src/inputoutput.jl#L7-L14">source</a></section><h3><a class="nav-anchor" id="Write-DCA-parameters-to-file.-1" href="#Write-DCA-parameters-to-file.-1">Write DCA parameters to file.</a></h3><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="DCATools.writeparam" href="#DCATools.writeparam"><code>DCATools.writeparam</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">function writeparam(outfile::String, g::DCAgraph; format=&quot;mat&quot;)</code></pre><p>Write graph <code>g</code> to file <code>outfile</code>: </p><ul><li>as a matrix if <code>format==&quot;mat&quot;</code></li><li>using <code>J i j a b value</code> if <code>format==&quot;mcmc&quot;</code></li></ul></div></div><a class="source-link" target="_blank" href="https://github.com/PierreBarrat/DCATools/blob/2c962efbb29e5d8f1db5365d439bae02d3b36753/src/inputoutput.jl#L98-L104">source</a></section><h3><a class="nav-anchor" id="Read-a-numerical-alignment-from-a-file.-1" href="#Read-a-numerical-alignment-from-a-file.-1">Read a numerical alignment from a file.</a></h3><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="DCATools.readmsanum" href="#DCATools.readmsanum"><code>DCATools.readmsanum</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">function readmsanum(infile::String ; format=1, header=false)</code></pre><p>Read an MSA stored in <code>infile</code> in a numerical format. </p><p>If <code>format=1</code>, amino acids should be mapped from 1 to <code>q</code>. If <code>format=0</code>, they should be mapped from 0 to <code>q-1</code>. <code>header</code> argument allows for discarding the first line of <code>infile</code>. </p></div></div><a class="source-link" target="_blank" href="https://github.com/PierreBarrat/DCATools/blob/2c962efbb29e5d8f1db5365d439bae02d3b36753/src/inputoutput.jl#L142-L149">source</a></section><h2><a class="nav-anchor" id="Model-tools-–-modeltools.jl-1" href="#Model-tools-–-modeltools.jl-1">Model tools – modeltools.jl</a></h2><h3><a class="nav-anchor" id="Switch-gauge-of-DCA-parameters.-1" href="#Switch-gauge-of-DCA-parameters.-1">Switch gauge of DCA parameters.</a></h3><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="DCATools.switchgauge!" href="#DCATools.switchgauge!"><code>DCATools.switchgauge!</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">switchgauge!(g::DCAgraph ; gauge=&quot;0sum&quot;)</code></pre><p>Switch parameters in <code>g</code> to gauge <code>gauge</code>.  Implemented gauges:</p><ol><li>0 sum: &quot;0sum&quot;</li><li>Lattice gas: &quot;LG&quot;, &quot;lg&quot;, &quot;latticegas&quot;. </li></ol></div></div><a class="source-link" target="_blank" href="https://github.com/PierreBarrat/DCATools/blob/2c962efbb29e5d8f1db5365d439bae02d3b36753/src/modeltools.jl#L3-L10">source</a></section><h3><a class="nav-anchor" id="Compute-energies-of-sequences-for-given-DCA-parameters.-1" href="#Compute-energies-of-sequences-for-given-DCA-parameters.-1">Compute energies of sequences for given DCA parameters.</a></h3><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="DCATools.computeenergies" href="#DCATools.computeenergies"><code>DCATools.computeenergies</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">computeenergies(g::DCAgraph, sample::Array{Int64,2})</code></pre><p>Compute energies of all configurations in <code>sample</code> with graph <code>g</code>.</p></div></div><a class="source-link" target="_blank" href="https://github.com/PierreBarrat/DCATools/blob/2c962efbb29e5d8f1db5365d439bae02d3b36753/src/modeltools.jl#L75-L79">source</a><div><div><pre><code class="language-none">computeenergies(g::DCAgraph, sample::Array{Int64,1})</code></pre><p>Compute energies of all configurations in <code>sample</code> with graph <code>g</code>.</p></div></div><a class="source-link" target="_blank" href="https://github.com/PierreBarrat/DCATools/blob/2c962efbb29e5d8f1db5365d439bae02d3b36753/src/modeltools.jl#L95-L99">source</a></section><h3><a class="nav-anchor" id="Infer-a-profile-model-from-frequencies-or-an-alignment.-1" href="#Infer-a-profile-model-from-frequencies-or-an-alignment.-1">Infer a profile model from frequencies or an alignment.</a></h3><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="DCATools.inferprofile" href="#DCATools.inferprofile"><code>DCATools.inferprofile</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">inferprofile(Y::Array{Int64,2}; q=findmax(Y)[1], pc = 1e-5, weights=[], save::String=&quot;&quot;)</code></pre><p>Infer profile model from alignment <code>Y</code>. </p><p>Keywords:</p><ul><li><code>q</code>: Default to maximum value in <code>Y</code>. </li></ul><p>-<code>pc</code> and <code>save</code>: See <code>inferprofile</code>. </p><ul><li><code>weights</code>: see <code>computefreqs</code></li><li></li></ul></div></div><a class="source-link" target="_blank" href="https://github.com/PierreBarrat/DCATools/blob/2c962efbb29e5d8f1db5365d439bae02d3b36753/src/modeltools.jl#L105-L115">source</a><div><div><pre><code class="language-none">inferprofile(f1::Array{Float64,1}, q::Int64; pc = 1e-5, save::String=&quot;&quot;)</code></pre><p>Infer profile model from frequencies <code>f1</code>. </p><p>Keywords: -<code>pc</code>: Pseudocount ratio. Defaults to <code>1e-5</code>.</p><ul><li>save: File to save inferred profile. Format of save is <code>mat</code>, readable by <code>readdlm</code> or <code>readparam</code>. </li></ul></div></div><a class="source-link" target="_blank" href="https://github.com/PierreBarrat/DCATools/blob/2c962efbb29e5d8f1db5365d439bae02d3b36753/src/modeltools.jl#L121-L129">source</a></section><h2><a class="nav-anchor" id="Contact-prediction-–-contactprediction.jl-1" href="#Contact-prediction-–-contactprediction.jl-1">Contact prediction – contactprediction.jl</a></h2><h3><a class="nav-anchor" id="Compute-Fapc-score-for-given-coupling-matrix.-1" href="#Compute-Fapc-score-for-given-coupling-matrix.-1">Compute Fapc score for given coupling matrix.</a></h3><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="DCATools.Fapc" href="#DCATools.Fapc"><code>DCATools.Fapc</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">Fapc(A::Array{Float64,2}, q::Int64 ; APC::Bool = true, gap::Bool = false, cols::Int64=3)</code></pre><p>Compute Frobenius norm of <code>q x q</code> blocks in matrix <code>A</code>. </p><p>Keywords:</p><ul><li><code>APC</code>: apply the famous APC correction. Default to <code>true</code>.</li><li><code>gap</code>: Remove the state <code>1</code> from the Frobenius norm. Default to <code>false</code>.</li><li><code>col</code>: Format of output. With <code>3</code>, output is an array with rows being <code>i j value</code>. With <code>1</code>, output is a vector containing <code>value</code> only. Default to <code>3</code>. </li></ul></div></div><a class="source-link" target="_blank" href="https://github.com/PierreBarrat/DCATools/blob/2c962efbb29e5d8f1db5365d439bae02d3b36753/src/contactprediction.jl#L47-L57">source</a></section><h3><a class="nav-anchor" id="Compute-PPV-for-given-scores-and-distances.-1" href="#Compute-PPV-for-given-scores-and-distances.-1">Compute PPV for given scores and distances.</a></h3><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="DCATools.PPV" href="#DCATools.PPV"><code>DCATools.PPV</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">PPV(scores::Array{Float64,2}, distances::Array{Float64,2} ; minrange=4, threshold=8)</code></pre><p>Compute Positive Predictive Value (PPV) for <code>scores</code> and <code>distances</code>. Both of these arrays should be in the format <code>i j val</code>. </p><p>Keywords:</p><ul><li><code>minrange</code>: Minimum value of <code>|i-j|</code> for a prediction to be made. Default to <code>4</code>.</li><li><code>threshold</code>: Threshold defining contact. Default to <code>8</code>. </li></ul></div></div><a class="source-link" target="_blank" href="https://github.com/PierreBarrat/DCATools/blob/2c962efbb29e5d8f1db5365d439bae02d3b36753/src/contactprediction.jl#L4-L12">source</a></section><footer><hr/></footer></article></body></html>
